Add string.is_match helper function

sorcerykid opened the pull request:
This PR adds a string library function for conditional pattern matching with capture groups stored to the underscore variable. The concept is intended to be similar to the way that Perl handles captures.

I often find myself re-implementing this functionality in various mods to better structure my code (less redundancy and duplication). Traditionally, this required cascading if/else blocks, which obfuscates the logic and results in extraneous variables. I feel it is suitable for the official Minetest API.

Possible use-cases:

- Serial if/elseif blocks in combination with multiple captures for input validation.
- Easier localization of capture variables within the scope of each conditional branch

This is an example of a snippet of code from my Bitwise Operators mod:

```
if option == "b" and
		(is_match(attrs, "^([0-]?)([0-9]*)$") or is_match(attrs, "^([0-]?)([0-9]*):([0-9]+)$")) then
	return parse_token(remove(args, 1), _[ 1 ], _[ 2 ], _[ 3 ])
else
	return string.format(token, remove(args, 1))
end
```

Without is_match(), the code expands from 6 to 15 lines, with significant repetition:

```
if option == "b" then
	local a1, a2 = string.match(attrs, "^([0-]?)([0-9]*)$")
	if a1 and a2 then
		parse_token(remove(args, 1), a1, a2)
	else
		local a1, a2, a3 = string.match(attrs, "^([0-]?)([0-9]*):([0-9]+)$")
		if a1 and a2 then
			parse_token(remove(args, 1), a1, a2, a3)
		else
			return string.format(token, remove(args, 1))
		end
	end
else
	return string.format(token, remove(args, 1))
end
```

## To do

This is my first PR on GitHub. So, I'm open to core dev feedback if there is anything I did wrong or that I need to fix. I really look forward to learning and contributing!

## How to test

Edit the `/sethome` chat command in Minetest Game as follows:

```
minetest.register_chatcommand("sethome", {
	description = "Set your home point",
	privs = {home = true},
	func = function(name,params)
		name = name or "" -- fallback to blank name if nil
		local player = minetest.get_player_by_name(name)

		if not player then
			return false, "Player not found!"
		end

		local pos = minetest.string_to_pos(params) or player:get_pos()
		sethome.set(name,pos)
		return true, "Home set!"
	end,
})
```

Then in singleplayer, issue the chat commands `/sethome 0 10 0` and `/sethome (5,0,5)`.

rubenwardy commented:
Underscore is used as a common notation for unused, so shouldn't be used as a return variable. Returning via global state is also very dodgy. Reduced lines of code is less important that readability and stability

A better way to do this would be to add support for c++ regex, which allows optional groups and captures. Or a pure lua way would be something like:

```lua
local matches = string.matches("input", { "pattern1", "pattern2", "pattern3" })
```

rubenwardy commented:
```lua
local function op_match(attrs)
	local a1, a2 = string.match(attrs, "^([0-]?)([0-9]*)$")
        if a1 then
		return a1, a2
	end

	local a1, a2, a3 = string.match(attrs, "^([0-]?)([0-9]*):([0-9]+)$")
        if a1 then
		return a1, a2, a3
	end
end

if option == "b" then
	local a1, a2, a3 = op_match(attrs)
	if a1 and a2 then
		parse_token(remove(args, 1), a1, a2, a3)
	else
		return string.format(token, remove(args, 1))
	end
else
	return string.format(token, remove(args, 1))
end
```

SmallJoker commented:
Please don't use global variables to store return values. This pollutes the environment and even permits mods to read out whatever was checked last.

As an alternative to @rubenwardy's code passing a table (by reference) is also possible:
```Lua
local captures = {}
if foo:is_match("(%s+)", captures) or
		foo:is_match("(%d+)", captures) then
	print(captures[1])
end
```

sorcerykid commented:
I appreciate all your input @Krock and @rubenwardy.

I will go ahead and close this since it's not feasible.

rubenwardy commented:
> As an alternative to @rubenwardy's code passing a table (by reference) is also possible:

I forgot that this was possible. That would make the function this:

```lua
function string.is_match(str, pattern, capture)
    local ret = table.pack(string.match(str, pattern))
    if capture then
        for i=1, #ret do
            capture[i] = ret[i]
        end
    end 
    return #ret > 0
end

sorcerykid commented:
Given the tiny size of the function, I think it's simpler to keep it local as I've been doing rather than integrating it into the string library. At the head of my mods I simply declare the function

```
local _

local function is_match( text, pattern )
     _ = { string.match( text, pattern ) }
     return #_ > 0
end
```

This addresses all the concerns and still allows for returning the underscore variable, since it's non-global,
